library/src/scala/tasty/util/ShowSourceCode.scala

```diff
@@ -0,0 +1,751 @@
+        this += "]"
+
+      case TypeTree.Annotated(tpt, annots) =>
+        printTypeTree(tpt)
+        // TODO print annots
+
+      case TypeTree.And(left, right) =>
+        printTypeTree(left)
+        this += " & "
+        printTypeTree(right)
+
+      case TypeTree.Or(left, right) =>
+        printTypeTree(left)
+        this += " | "
+        printTypeTree(right)
+
+      case TypeTree.ByName(result) =>
+        this += "=> "
+        printTypeTree(result)
+
+    }
+
+    def printTypeOrBound(tpe: TypeOrBounds): Buffer = tpe match {
+      case tpe@TypeBounds(lo, hi) =>
+        this += " >: "
+        printType(lo)
+        this += " <: "
+        printType(hi)
+      case tpe@Type() => printType(tpe)
+    }
+
+    def printType(tpe: Type): Buffer = tpe match {
+      case Type.ConstantType(const) =>
+        printConstant(const)
+
+      case Type.SymRef(sym, prefix) =>
+        prefix match {
+          case EmptyPackage() =>
+          case RootPackage() =>
+          case prefix@Type.SymRef(ClassDef(_, _, _, _, _), _) =>
+            printType(prefix)
+            this += "#"
+          case prefix@Type() =>
+            printType(prefix)
+            this += "."
+          case prefix@NoPrefix() =>
+        }
+        printDefinitionName(sym)
+
+      case Type.TermRef(name, prefix) =>
+        prefix match {
+          case prefix@Type() =>
+            printType(prefix)
+            if (name != "package")
+              this += "." += name
+            this
+          case NoPrefix() =>
+            this += name
+        }
+
+      case Type.TypeRef(name, prefix) =>
+        prefix match {
+          case NoPrefix() =>
+          case prefix@Type() =>
+            printType(prefix)
+            this += "."
+        }
+        this += name.stripSuffix("$")
+
+      case Type.SuperType(thistpe, supertpe) =>
+        ???
+
+      case Type.Refinement(parent, name, info) =>
+        printType(parent)
+        // TODO add refinements
+
+      case Type.AppliedType(tp, args) =>
+        printType(tp)
+        this += "["
+        printTypesOrBounds(args, ", ")
+        this += "]"
+
+      case Type.AnnotatedType(tp, annot) =>
+        printType(tp)
+
+      case Type.AndType(left, right) =>
+        printType(left)
+        this += " & "
+        printType(right)
+
+      case Type.OrType(left, right) =>
+        printType(left)
+        this += " | "
+        printType(right)
+
+      case Type.ByNameType(tp) =>
+        this += " => "
+        printType(tp)
+
+      case Type.ParamRef(binder, idx) =>
+        ???
+
+      case Type.ThisType(tp) =>
+        printType(tp)
+
+      case Type.RecursiveThis(rec) =>
+        ???
+
+      case Type.RecursiveType(tp) =>
+        ???
+
+      case Type.MethodType(paramNames, paramTypes, resType) =>
+        ???
+
+      case Type.PolyType(paramNames, paramBounds, resType) =>
+        ???
+
+      case Type.TypeLambda(paramNames, paramBounds, resType) =>
+        ???
+    }
```

Comment:
It seems `???` in the above means they should never be reached. Maybe use a "|" pattern to combine them in one case (the variable patterns are not used, can be underscore), and add a note.

Reply:
I will add a better way to handle unreachable patterns.